include/mbedtls/config.h

```diff
@@ -1520,8 +1520,8 @@
 
 /** \def MBEDTLS_SSL_EXTENDED_MASTER_SECRET
  *
- * Enable support for Extended Master Secret, aka Session Hash
- * (draft-ietf-tls-session-hash-02).
+ * Enable support for RFC 7627: Session Hash and Extended Master Secret
+ * Extension.
```

Comment:
I checked the [differences between the draft and the RFC](https://tools.ietf.org/rfcdiff?url1=https://tools.ietf.org/id/draft-ietf-tls-session-hash-02.txt&url2=https://tools.ietf.org/rfc/rfc7627.txt) and found the following potential differences:

* :heavy_check_mark: SSL 3.0 now forbidden: we never implemented SSLv3 support described by the draft RFC.
* :heavy_check_mark: Possibility of using the feature with DTLS: we do that.
* :information_source: An endpoint that mandates extended master secret must abort if the peer does not declare the extension: not relevant because we don't support marking the extension as mandatory, only optional or disabled.
* :information_source: “The client SHOULD NOT offer an abbreviated handshake to resume a session that does not use an extended master secret.” We don't base decisions about resumption on the presence of the EMS extension.
* :question: “If the original session used the "extended_master_secret" extension but the new ClientHello does not contain it, the server MUST abort the abbreviated handshake.” Similar provision on the client side. We don't check this explicitly. However, isn't this provision redundant since the extension changes the resulting master secret? We don't test this.

Reply:
> However, isn't this provision redundant since the extension changes the resulting master secret?
There's no new master secret for abbreviated handshakes. (But if you had in mind the case of an active attacker stripping the ClientHello from this extensions - first it will be detected as the Finished values won't match, and second I don't see what an attacker could gain from that considering no new MS is getting computed...)

I think this provision is motivated by the general principle that when a client offers resumption, it should also offer everything it offered in the original handshake - in particular so that if the server can't or doesn't want to resume, the same security parameters as in the original handshake are going to be used.

I think the RFC is just taking a hard stance on this principle here. (Also, there's no good reason for a client that used EMS in the past to stop using it - in general the server can't track that, but in the case of resumption it can, so we might as well check for it.)

So I think this is only a formal compliance issue with no actual consequence. We could however solve that, as well as improve on the previous point, by storing in `ssl_session` a bit indicated whether EMS was negociated, and then check that bit when resuming.

(I'd also consider adding an option to enforce the use of the extension of nice "enhancement" item to add to our backlog whenever we get some time - could also be a good first/second contirbution for someone gaining familiarity with TLS.)

Wdyt?

Reply:
I was thinking of a vulnerability that would allow an attacker to do something with non-EMS connections, but that EMS would protect against. I didn't have a specific example of a vulnerability in mind. Could there be a downgrade attack where the attacker leverages a weakness in the handshake to resume an EMS session, which only arises when the handshake negotiation is for a non-EMS session (at least as seen by one of the sides)?

Reply:
I don't see how an attacker could send a correct Finished message for an abbreviated handshake unless they already managed to compromise the MS of the session being resumed - in which case it's game over anyway.